### PR TITLE
feat: [M12] DATA-8: the tool-calling and reasoning SFT corpora have no training driver

### LINKED ISSUE #306: [M12] DATA-8: the tool-calling and reasoning SFT corpora have no training driver
Found by `/closure-audit` (2026-08-18), pass A. Capability row `DATA-8` in
`docs/feature-matrix.md`.

## Capability
A person can run an SFT pass over the tool-calling and reasoning corpora this repo already knows
how to build — without hand-converting them into the generic format first.

## The gap

Two corpus builders exist, are fully tested, and have **no non-test consumer**:

- `src/data/tool_sft.py` — `build_tool_sft` plus a `__main__` CLI (`:142,229,254`). Only
  `tests/test_tool_sft.py` touches it.
- `src/data/reasoning_sft.py` — `build_reasoning_sft` (`:73,167,195`). Only
  `tests/test_reasoning_sft.py` and `tests/test_storage.py` touch it.

`scripts/sft.py` reads only generic `data/sft/{train,val}.jsonl`; it has no handling for the
`reasoning-packed/` form or for tool-call rows. So both builders produce output nothing consumes.

This is self-documented in two places, which is how it stayed invisible:

- `docs/reserve/path-b-run.md:239-242` — *"`src/data/reasoning_sft.py` builds the data, but there
  is **no dedicated reasoning-SFT training driver** in `scripts/`"*.
- `docs/design/11-post-training.md:56-62` — a **live, non-reserve** doc describing the two output
  forms as though they are usable.

#102 ("Tool-use SFT") was closed COMPLETED with only the builder landed. #101 does not cover this:
its body is M10/#65 framing about an instruct layer and a conversion teacher, not these corpora
(see the comment left on #101).

## Scope

This is one vertical slice, not two, because both failures are the same seam: `scripts/sft.py`
cannot consume a corpus form that is not the generic one.

In scope:
- Teach `scripts/sft.py` to consume the tool-call and reasoning-packed forms the builders emit —
  a corpus-form argument or an autodetect from the manifest, whichever matches the existing
  loader conventions.
- Keep response masking correct for each form: the assistant span for reasoning traces, the
  tool-call payload for tool rows.
- One end-to-end test per form, from builder output through a training step.

Out of scope:
- New corpus sources, new data, or changing what the builders emit.
- DPO/GRPO over these corpora.
- Re-framing #101, which is its own decision.

## Definition of done
- [x] `scripts/sft.py` trains from `build_tool_sft` output without manual conversion.
- [x] `scripts/sft.py` trains from `build_reasoning_sft` output, including the `reasoning-packed/`
      form named in `docs/design/11-post-training.md`.
- [x] Response masking is correct per form and asserted by a test — a wrong mask trains on the
      prompt and is silent.
- [x] Chat-template EOS stays consistent across SFT, RL and serving; a mismatch degrades the model
      at serving time and is the failure this area is most prone to.
- [x] Both builders now have a real non-test consumer, which is the condition that closes this.
- [x] `docs/design/11-post-training.md` describes how to actually run each form.
- [x] `docs/reserve/path-b-run.md:239-242`'s "no dedicated driver" statement is corrected — it is
      a reserve doc, but that specific sentence becomes false.
- [x] `docs/feature-matrix.md` `DATA-8` moves to `Shipped`; `docs/test-plan.md`'s `DATA-8` Gaps
      cell is rewritten.

## Edge cases
- A reasoning trace longer than the sequence length → truncation must not orphan the answer span.
- A tool row with no tool call, or with several.
- Malformed JSON in a tool-call payload → a named error, not a crash mid-epoch.
- An empty corpus after filtering → fail loudly rather than train on nothing.
- Mixing forms in one run → either supported and tested, or rejected with a clear message.

## Verification
- `.venv/bin/python -m pytest tests/test_tool_sft.py tests/test_reasoning_sft.py tests/test_sft_train_step.py tests/test_sft_loader.py -q`
- `python -m src.data.tool_sft --out <tmp>` then `.venv/bin/python scripts/sft.py --config config/toy.yaml --data <tmp>` completing a step.
- The same for `build_reasoning_sft`.
- `.venv/bin/python -m pytest -q` full suite green.

### PR DESCRIPTION
**Pull request by travisgalloway**
Closes #306

## Summary

`build_tool_sft`, `build_reasoning_sft` and `build_instruct_sft` were fully built, fully tested, and had **no non-test consumer**: they write one masked-JSONL file per form plus a manifest under `shared/sft/tokenized/<tok>-<k>/`, while `scripts/sft.py` hard-coded `<data>/train.jsonl` + `<data>/val.jsonl`. Nothing bridged the two.

**No new trainer was added.** `scripts/sft.py` already wires the whole M9 path (`SFTLoader` → `make_sft_train_step` → `train.loop` → `CheckpointStore` → `evaluate_masked`, with grad accum, fp16 scaling, MoE balancer, resume and `--backend`). This is a corpus resolver plus wiring; `git diff --stat main -- scripts/` touches `scripts/sft.py` only.

- **`src/data/sft_corpus.py`** (new, portable) — resolves `--data` to either the generic pair of paths *untouched* (every pre-#306 invocation is bit-for-bit unchanged: no split, no filter, no manifest check) or an in-memory train/val split over the requested `shared/sft` forms.
- **`scripts/sft.py`** — `--corpus-form {instruct,reasoning,tool,reasoning-packed,generic,auto}` (several may be mixed), `--val-frac`, `--max-len`, and a runbook in the docstring. **Nothing below the loader construction changed.**
- **`src/data/sft_loader.py`** — `load_sft_records()` names the file and line on a malformed line, and rejects a record whose three field lengths disagree (`_collate` right-pads each field independently, so a short `loss_mask` silently shifts off its tokens and trains the prompt). `SFTLoader(records=...)` takes the resolver's split without a temp file.

### The packed decision

`reasoning-packed/` is **rejected by name**, printing the recipe that works. It is `shard.pack_atomic` output: `.bounds` marks document *starts* only, every document is padded to `chunk_align` and every sequence tail is `pad_id`, so nothing separates a real final token from padding and **no loss mask can be reconstructed**. An "SFT over packed" mode would have to invent one. Packed is a pretraining artifact and gets the pretraining driver — and that path is now proved by a test, not just asserted: `split_shards` → `PackedLoader` → `make_train_step`, one update that moves the weights.

```
$ .venv/bin/python scripts/sft.py --config config/toy.yaml --data <tok_dir> --corpus-form reasoning-packed ...
ValueError: --corpus-form reasoning-packed: not an SFT form. `reasoning-packed/` carries no loss
mask, so it is a PRETRAINING artifact and scripts/sft.py cannot train it. Use the pretraining driver:
    python -m src.data.split --shards <...>/reasoning-packed --out <split> --val-tokens N
    python scripts/train.py --config <cfg> --data <split>
See docs/design/11-post-training.md (Thinking).
```

### Guards that now fire (each watched failing before it was kept)

- **chat EOS is verified at load, not merely recorded.** `chat_eos != chat_template.CHAT_EOS`, `template != "qwen-chatml"`, **or a missing `chat_eos` key** (unknown provenance is not a pass) → `ValueError` naming the file, expected and found. Watched red by disabling `validate_manifest` and confirming all three tests failed, then restored.
- **Form mixing** is supported when `chat_eos`/`template`/`tokenizer`/`model_id`/`seq_len` agree, refused by name when they do not.
- **Over-length records are dropped and counted, never truncated** — truncating clips the answer span off the end of the mask. Survivors' masks are asserted byte-identical to the builder's.
- **Empty corpus**, corpus emptied by the length filter, and too-few-to-split each get a distinct named error. Never a run on nothing.

## Testing

- [x] Tests added/updated — `tests/test_sft_corpus.py` (32 unit/integration, portable) and `tests/test_sft_driver_e2e.py` (7 e2e, MLX-gated)
- [x] All tests passing — `.venv/bin/python -m pytest -q -rs` → **1763 passed, 22 skipped in 377.81s** (no `test_parity_fixture_export.py` #298 flake this run)
- [x] Manual testing completed

**The anti-thin-wrapper test.** `tests/test_sft_driver_e2e.py` runs `scripts/sft.py:main()` in-process over real builder output: 20 real optimizer steps, every logged loss finite, final loss strictly below the first, and `<out>/resume` reloading to the same step with **byte-identical** weights — for the tool form, the reasoning form, and the two mixed. The mask is checked where it matters: a batch drawn from the *resolved loader* is compared against `chat_template.response_spans` recomputed from the cleaned rows, with the assistant span (through its trailing `<|im_end|>`) at 1 and system / user / `<tool_response>` / padding at 0. And the signal is shown to depend on it — an all-zero mask gives `loss == grad_norm == 0.0`, so a mask that had silently covered the prompt could not have produced the passing result.

**CI cost.** The whole new suite is **0.89 s** on this machine (39 tests), against the ~30 s budget — toy config, byte fallback, the checked-in handauthored sources. No CI job added; `tests/test_workflow_triggers.py` passes unedited and `git diff --stat main -- .github tests/test_workflow_triggers.py` is empty.

**Smoke gate** (fresh toy split): `SMOKE TEST PASSED ✅ resume is exact and eval runs.`

<details>
<summary>Manual verification (the issue's own steps)</summary>

```
$ python -m src.data.tool_sft --sources handauthored --byte-fallback --out-root /tmp/s306 \
      --seq-len 1024 --max-seq-len 1024
tool sft: 3 records (0 skipped, 0 schema-invalid, 2361 tokens, 2 abstention, 3 with distractors) -> .../tool.jsonl

$ python -m src.data.reasoning_sft --sources handauthored --byte-fallback --out-root /tmp/s306 \
      --seq-len 1024 --chunk-align 64
reasoning sft: 4 masked records (0 over-length dropped, 903 tokens), packed 4 traces atomically x1024

$ .venv/bin/python scripts/sft.py --config config/toy.yaml \
      --data /tmp/s306/sft/tokenized/qwen3-1k --corpus-form tool --max-len 1024 \
      --init /tmp/s306/base.safetensors --out /tmp/s306/run \
      --total-steps 20 --batch-size 2 --grad-accum 1 --base-lr 1e-3
[corpus] forms=tool train=2 val=1 dropped_overlength=0
[init] from pretrained base /tmp/s306/base.safetensors
[sft] params~0.1M  examples=2  total_steps=20  warmup=1  precision=fp32
step=0   lr=0.001      loss=7.743  grad_norm=1.984  val_loss=7.776  val_perplexity=2382
step=10  lr=0.0005872  loss=2.397  grad_norm=0.988  val_loss=4.284  val_perplexity=72.55
[done] step=20  val_loss=4.2452  val_perplexity=69.7693
```

(`--max-len` defaults to the config's `seq_len`; toy's is 128, so the recipe passes `1024` to keep the handauthored records — the default would otherwise refuse the run by name rather than silently drop everything.)

</details>

## Docs updated in the same commits

- `docs/design/11-post-training.md` — a runnable recipe per form, the two-driver split for reasoning (masked JSONL → `sft.py`; packed → `split --shards` → `train.py`) **and why**, a new "Running an SFT pass over the shared corpora" section with the `--data`/`--corpus-form` table and the split/over-length/mixing rules, and the note that `chat_eos` is now verified at load.
- `docs/reserve/path-b-run.md` — the "no dedicated reasoning-SFT training driver" bullet corrected; the never-valid `--data .../instruct` path (a directory that the builder never wrote — it writes `instruct.jsonl`) fixed to the tokenized dir plus `--corpus-form instruct`; the tool-SFT "code not built" bullet corrected.
- `docs/feature-matrix.md` — `DATA-8` → **Shipped**, capability text now "...and train on it".
- `docs/test-plan.md` — `DATA-8` gains `test_sft_corpus.py` (Integration) and `test_sft_driver_e2e.py` (E2E), the eight edge cases, and a Gaps cell that no longer says "no driver at all".

## Parked (4, in `docs/parked-findings.md`, none acted on)

`tool_sft._iter_calls` silently skips a syntactically malformed `<tool_call>` payload rather than counting it in `n_schema_invalid` (pinned by a test that records the observed behaviour); `PackedLoader`'s `seq_len+1` stride re-straddles `pack_atomic`'s atomic documents; `instruct_sft` writes `manifest.json` while the other two write `<kind>-manifest.json`; `scripts/dpo.py` / `scripts/rlvr.py` still have no path to these corpora.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_011ur4BdaFTqnwqSUAx5TrX8

### CODE REVIEW

File: src/data/sft_loader.py [hunk 1]
```diff
@@ -22,10 +22,52 @@
...
+            if not isinstance(rec, dict) or any(k not in rec for k in _REQUIRED_KEYS):
+                missing = [k for k in _REQUIRED_KEYS
+                           if not isinstance(rec, dict) or k not in rec]
+                raise ValueError(f"{path}:{lineno}: SFT record is missing {missing} "
+                                 f"(expected {list(_REQUIRED_KEYS)})")
+            lengths = {k: len(rec[k]) for k in _REQUIRED_KEYS}
+            if len(set(lengths.values())) != 1:
+                raise ValueError(f"{path}:{lineno}: SFT record field lengths disagree "
+                                 f"({lengths}) — the loss mask would not line up with its tokens.")
+            records.append(rec)
```
[the remark this answers is not in the document]
Comment (travisgalloway) [reply src/data/sft_loader.py:69]: ✅ Fixed in 0c4ae59cda7369a8cb977b0d782465eef71ee7bb — `load_sft_records()` now validates each required field is a list of numeric elements before computing lengths, raising a named `path:line` error (matching the existing contract) instead of letting a non-list/non-numeric value blow up with a bare `TypeError` inside `len()`/collation.

File: scripts/sft.py [hunk 2]
```diff
@@ -31,11 +61,33 @@
...
+                         "forms may be mixed. 'reasoning-packed' is rejected by name (no loss "
+                         "mask — see the module docstring).")
+    ap.add_argument("--val-frac", type=float, default=0.05,
+                    help="held-out fraction when the corpus has no val.jsonl of its own "
+                         "(the shared/sft forms never do); split is seeded by --seed")
+    ap.add_argument("--max-len", type=int, default=None,
+                    help="drop records longer than this, never truncate (0 = keep all; "
+                         "default: the config's seq_len). Applies to the shared/sft layout "
+                         "only — a generic train.jsonl/val.jsonl pair is passed through "
+                         "untouched.")
```
[the remark this answers is not in the document]
Comment (travisgalloway) [reply scripts/sft.py]: ✅ Fixed in 0c4ae59cda7369a8cb977b0d782465eef71ee7bb — updated the `--max-len` help text. It was wrong: `resolve_sft_corpus()` also applies `max_len` when a generic corpus is missing `val.jsonl` (it falls through to the same `_split_to_corpus()` path shared/sft forms use). Only a generic `train.jsonl`+`val.jsonl` pair is passed through untouched — the help text now says so.

File: src/data/sft_corpus.py [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+                     data_dir: Path) -> SFTCorpus:
+    """Length-filter, fail loudly on empty, then split deterministically."""
+    n_raw = len(records)
+    if n_raw == 0:
+        raise ValueError(f"empty SFT corpus: {data_dir} yielded 0 records for forms "
+                         f"{'+'.join(names)}. Nothing to train on.")
+
+    n_dropped = 0
+    if max_len:
+        kept = [r for r in records if len(r["input_ids"]) <= max_len]
```
[the remark this answers is not in the document]
Comment (travisgalloway) [reply src/data/sft_corpus.py:276]: ✅ Fixed in 0c4ae59cda7369a8cb977b0d782465eef71ee7bb — `_split_to_corpus()` now rejects `max_len < 0` explicitly up front (`--max-len must be >= 0 (0 = keep all), got {max_len}`) instead of letting the negative value pass the truthy check and empty the corpus with a confusing "longer than max_len=-1" error.
